### ISSUE TITLE
JPEG2000KakaduImageReader causes JVM to crash For Certain JP2 images

### ISSUE THREAD

**Issue by bbarberBPL** [cantaloupe-project/cantaloupe#303]
The JPEG2000KakaduImageReader causes the JVM and produces a core dump log file.  This does not occur with all images.  The following error occurs in the logs right before the JVM crashes

The errors are similar to [Issue 300](https://github.com/cantaloupe-project/cantaloupe/issues/300)

`ERROR [qtp920011586-1304] e.i.l.c.p.c.JPEG2000KakaduImageReader [JPEG2000KakaduImageReader.java:157] Kakadu Core Error:
You may not apply restrictions to the resolution or number of image components
after the first tile access, unless the codestream object is set up to be
persistent.`

Here is the core dump file.
[hs_err_pid2152.log](https://github.com/cantaloupe-project/cantaloupe/files/3462341/hs_err_pid2152.log)

Also here are the the test images as well as some log snippets for testing.
[cantaloupe-413-badimg2.zip](https://github.com/cantaloupe-project/cantaloupe/files/3462350/cantaloupe-413-badimg2.zip)

**Comment by adolski** [comment]
Hi @bbarberBPL,

I can't reproduce this. Are you sure that the images in the zip file are the ones associated with the crashes?

For example, one of the errors has thread ID `qtp920011586-58.` Did this thread handle other requests after the one for `/commonwealth:cn69mp97w/full/,500/0/default.jpg`?

**Comment by bbarberBPL** [comment]
Actually I have a more recent one for you
the url it was hitting was `/full/pct:30/0/default.jpg`

[bad_img.jp2.zip](https://github.com/cantaloupe-project/cantaloupe/files/3462760/bad_img.jp2.zip)

**Comment by adolski** [comment]
That one works for me too.

**Comment by bbarberBPL** [comment]
I'm not sure what happened with those initial ones. It could have been how I was grepping the logs. I'll try and find more before the end of the day here. Otherwise I can give them to you Monday

**Comment by adolski** [comment]
Okay, thanks. I'll be out next week so take your time.

**Comment by bbarberBPL** [comment]
@adolski  So we are having issues reproducing the issues on our end as well. However I think i might have narrowed down the cause. 

In the core dump file the stack shows the following 

```Stack: [0x00007fd47626e000,0x00007fd47636f000],  sp=0x00007fd47636c4c0,  free space=1017k
Native frames: (J=compiled Java code, A=aot compiled Java code, j=interpreted, Vv=VM code, C=native code)
C  [libkdu_jni.so+0x2345ca]  kdu_supp::kdu_region_decompressor::process(int*, kdu_core::kdu_coords, int, int, int, kdu_core::kdu_dims&, kdu_core::kdu_dims&)+0xda

Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
J 6564  kdu_jni.Kdu_region_decompressor.Process([ILkdu_jni/Kdu_coords;IIILkdu_jni/Kdu_dims;Lkdu_jni/Kdu_dims;)Z (0 bytes) @ 0x00007fd5a8a69ae3 [0x00007fd5a8a69a40+0x00000000000000a3]
J 5154 c2 edu.illinois.library.cantaloupe.processor.codec.JPEG2000KakaduImageReader.readRegion(Ledu/illinois/library/cantaloupe/image/Rectangle;Ledu/illinois/library/cantaloupe/operation/Scale;Ledu/illinois/library/cantaloupe/image/ScaleConstraint;Ledu/illinois/library/cantaloupe/operation/ReductionFactor;[D)Ljava/awt/image/BufferedImage; (860 bytes) @ 0x00007fd5a8913d04 [0x00007fd5a89125a0+0x0000000000001764]
J 9771 c1 edu.illinois.library.cantaloupe.processor.KakaduNativeProcessor.process(Ledu/illinois/library/cantaloupe/operation/OperationList;Ledu/illinois/library/cantaloupe/image/Info;Ljava/io/OutputStream;)V (98 bytes) @ 0x00007fd5a221b514 [0x00007fd5a221acc0+0x0000000000000854]
J 10605 c1 edu.illinois.library.cantaloupe.resource.ImageRepresentation.process(Ljava/io/OutputStream;)V (65 bytes) @ 0x00007fd5a17cc23c [0x00007fd5a17cbfe0+0x000000000000025c]
J 9770 c1 edu.illinois.library.cantaloupe.resource.ImageRepresentation.copyOrProcess(Ljava/io/OutputStream;)V (55 bytes) @ 0x00007fd5a1d800e4 [0x00007fd5a1d7faa0+0x0000000000000644]
J 9985 c1 edu.illinois.library.cantaloupe.resource.ImageRepresentation.write(Ljava/io/OutputStream;)V (324 bytes) @ 0x00007fd5a229b544 [0x00007fd5a2299000+0x0000000000002544]
J 8473 c1 edu.illinois.library.cantaloupe.resource.iiif.v2.ImageResource.doGET()V (818 bytes) @ 0x00007fd5a2037c5c [0x00007fd5a2031320+0x000000000000693c]
```
The only place I see that `kdu_jni.Kdu_region_decompressor.Process` function call is [here](https://github.com/cantaloupe-project/cantaloupe/blob/98b962ce58693ba1953eff4ee8d0317bbe214d77/src/main/java/edu/illinois/library/cantaloupe/processor/codec/JPEG2000KakaduImageReader.java#L574) 

Also after digging around i found a cpp source file of the kakadu error we have been seeing that is causing the crash which is currently 
`You may not apply restrictions to the resolution or number of image components
after the first tile access, unless the codestream object is set up to be
persistent.`

The link to that file is [here](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&ved=2ahUKEwjxptffi_HjAhUCjVkKHbFaApUQFjAAegQIARAB&url=http%3A%2F%2Fextras.springer.com%2F2002%2F978-1-4613-5245-7%2FSoftware%2FKakadu_V2.2.3%2Fcoresys%2Fcompressed%2Fcodestream.cpp&usg=AOvVaw1YBeB6Uya97cE577xLiT5C) however it is an older version of the file and i'm not sure if the current one throws the error message the same way.  


This is the snippet in question from that file 
```{
  if (state->in == NULL)
    { kdu_error e; e << "Input restrictions may be applied only to codestream "
      "objects opened for input!"; }
  if (state->tiles_accessed)
    {
      if (state->num_open_tiles != 0)
        { kdu_error e; e << "You may apply restrictions to the resolution "
          "or number of image components only after closing all open tiles."; }
      if (state->tiles_accessed && !state->persistent)
        { kdu_error e; e << "You may not apply restrictions to the resolution "
          "or number of image components after the first tile access, unless "
          "the codestream object is set up to be persistent."; }
    }
```
Note that `state` is a pointer to a `kd_codestream`

All that being said I'm guessing possible this is some kind of threading issue. It seems like multiple threads are trying to access the same file in memory with different restrictions. This would explain why the code stream thinks the number of tiles open are not equal to  0. 

Another reason I think this is our production server has 6 cores(6 JVM threads) and our dev server(which we have not been able to reproduce this on) is only 1 core. 

I hope this helps

**Comment by ebenenglish** [comment]
Hi @adolski just following up a bit more...

We've been unable to replicate the errors we're seeing in Production in our Dev and Test environments, which leads me to believe the issue may not be with the files themselves.

I'm wondering if this is a high traffic/load problem perhaps? We've started to see these errors showing up in bunches in our logs, e.g.:
```
2019-08-12 15:17:15,772 ERROR [qtp920011586-16] e.i.l.c.p.c.JPEG2000KakaduImageReader [JPEG2000KakaduImageReader.java:157] Kakadu Core Error:
You may not apply restrictions to the resolution or number of image components
after the first tile access, unless the codestream object is set up to be
persistent.
2019-08-12 15:17:15,780 ERROR [qtp920011586-50] e.i.l.c.p.c.JPEG2000KakaduImageReader [JPEG2000KakaduImageReader.java:157] Kakadu Core Error:
You may not apply restrictions to the resolution or number of image components
after the first tile access, unless the codestream object is set up to be
persistent.
2019-08-12 15:17:15,910 ERROR [qtp920011586-12] e.i.l.c.p.c.JPEG2000KakaduImageReader [JPEG2000KakaduImageReader.java:157] Kakadu Core Error:
You may not apply restrictions to the resolution or number of image components
after the first tile access, unless the codestream object is set up to be
persistent.
```
When looking at the logs a bit further, it seems these were all requests for the same image (using the thread IDs to try and figure out which requests are resulting in errors):
```
2019-08-12 15:17:15,143 INFO [qtp920011586-16] e.i.l.c.r.i.v.ImageResource [AbstractResource.java:111] Handling GET /iiif/2/commonwealth:n009x597p/4096,0,1952,2048/976,/0/default.jpg
2019-08-12 15:17:15,146 INFO [qtp920011586-50] e.i.l.c.r.i.v.ImageResource [AbstractResource.java:111] Handling GET /iiif/2/commonwealth:n009x597p/4096,4096,1952,726/976,/0/default.jpg
2019-08-12 15:17:15,143 INFO [qtp920011586-12] e.i.l.c.r.i.v.ImageResource [AbstractResource.java:111] Handling GET /iiif/2/commonwealth:n009x597p/4096,2048,1952,2048/976,/0/default.jpg
```
Hope this helps!

**Comment by adolski** [comment]
Hi @bbarberBPL & @ebenenglish,

Thanks for looking into this, and sorry for the delay.

Kdu_region_decompressor is the main object involved in coordinating the decompression, and in its simplest usage, it's single-threaded. Cantaloupe uses its various method overloads to enable multi-threaded decompression, [assigning one thread per CPU core](https://github.com/cantaloupe-project/cantaloupe/blob/50560a6f97239ea984cdcc0bc7dd92419a80eb8a/src/main/java/edu/illinois/library/cantaloupe/processor/codec/JPEG2000KakaduImageReader.java#L389-L393).

I don't know where a thread-safety issue would be coming into play on the Cantaloupe side, as JPEG2000KakaduImageReaders aren't shared across threads. It's possible that there is some kind of issue in the Kdu_region_decompressor or elsewhere in Kakadu. But, I've run KakaduNativeProcessor on multiple cores all the time and never seen or heard of this myself.

So, I still don't know the root cause, but I have a couple of ideas for workarounds:

1. Make the number of threads used by Kdu_region_decompressor customizable. Then, it could be set to 1.
2. Since it looks like this error wouldn't occur with a persistent codestream, set the codestream as persistent. But I don't know what implications that might have for memory usage.

Would you be willing to test one or both of these on a separate branch off of `release/4.1`?

**Comment by bbarberBPL** [comment]
@adolski  No worries about the delay.

I think we should try the code stream persistence one first. Seems like it would be the easiest one. To clarify it seemed to be happening whenever we would get multiple requests at the same time for the same image across threads but within different regions. At least that's what I got from the logs. Since we are using the cache strategy for the source image I thought it might be the fact multiple threads were trying to access the same file when rendering different regions. It could also be an issue with kakadu itself so maybe adjusting the thread count would also help. Let me know when you have the persistent one working and I'll help you test that.

**Comment by adolski** [comment]
On the `github-303` branch there are two new configuration keys available. They can be used like:

```
KakaduNativeProcessor.num_threads = 1
KakaduNativeProcessor.persistent_codestream = true
```

I'll be interested to know if either or both have any effect.

**Comment by ebenenglish** [comment]
Hi @adolski sorry for the delay in replying, been out of the office quite a bit over the past week or two.

We'll try and test the `github-303` branch sometime next week, will report back. Thanks again!

**Comment by ogloke** [comment]
Is there a fix for this issue? It has been happening constantly to our cantaloupe server. I have added those keys to the config but it has made no difference.

**Comment by adolski** [comment]
@ogloke If not already using the latest Cantaloupe version, please upgrade to that first and then post the following:

1. the source JP2
2. the URI path that causes the crash, e.g. `/iiif/2/...`

**Comment by ogloke** [comment]
@adolski Hi, We are already running version 4.1.4. The crash doesn't seem to be related to any single image but seems to happen randomly. I have tried replicating the issue with images that were in use during crashes but could not. I have included here the image from the last crash, the URI and the error log
URI: https://trin-digital-library.trin.cam.ac.uk/iiif/2/O.3.7%2F067_O.3.7_f031r.jp2
[hs_err_pid9652.log](https://github.com/cantaloupe-project/cantaloupe/files/3754360/hs_err_pid9652.log)
[067_O.3.7_f031r.zip](https://github.com/cantaloupe-project/cantaloupe/files/3754365/067_O.3.7_f031r.zip)

**Comment by ogloke** [comment]
@adolski 
URI of image: https://trin-digital-library.trin.cam.ac.uk/iiif/2/O.3.7%2F067_O.3.7_f031r.jp2/full/full/0/default.jpg

**Comment by ogloke** [comment]
**Update:** disabling the derivative cache seems to have stabilised things

**Comment by ksclarke** [comment]
We're having the same problem. We're using 4.1.5. We see the same non-repeatability of the issue with the same image on a different server (i.e., we can move the "problematic image" from our prod server to a test server and the image is served fine).

We use containers, so our environments are the same. We're using JPXes and they were created with Kakadu and look fine to me. We're using an S3 bucket for the image source (different s3 buckets for test and prod) and another S3 bucket for the derivatives cache (different for test and prod).

Like @ogloke determined, the problem goes away if we disable the derivative cache. With it enabled, we consistently have Cantaloupe crashing and the container being restarted/self-healed.

We're still investigating but thought I'd post in case any of the others with the problem figured out the cause and just didn't update here (or found better workarounds).

**Comment by ksclarke** [comment]
Clearing that one derivative cache image and then trying to access it (regenerating it) also allowed us to see view that image (we didn't get the problem again on reviewing the image). This wasn't isolated to one image, though; we were seeing the crash on a bunch of JPX images. We've now wiped the whole cache and are going to turn the derivatives caching back on and try reproducing the issue from a clean cache.

**Comment by mannu2612** [comment]
Do we have a fix for this? I am using latest tag v4.1.6. I can't seem to find keys  `KakaduNativeProcessor.num_threads = 1
KakaduNativeProcessor.persistent_codestream = true` in `github-303` branch https://github.com/cantaloupe-project/cantaloupe/blob/github-303/src/main/java/edu/illinois/library/cantaloupe/config/Key.java

**Comment by mannu2612** [comment]
@ksclarke Were you able to solve this?

**Comment by adolski** [comment]
@mannu2612 Those keys are only on the `github-303` branch, but they aren't present within `Key` yet: https://github.com/cantaloupe-project/cantaloupe/blob/github-303/src/main/java/edu/illinois/library/cantaloupe/processor/KakaduNativeProcessor.java#L158-L164

Also, `github-303` was getting pretty stale, so I just merged `release/4.1` into it.

If you still notice this problem on `github-303`, the `develop` branch (version 5.0 in progress) now includes a newer version of the Kakadu library, which is also worth trying.

**Comment by mannu2612** [comment]
@adolski I see that we now have version 3 of iiif APIs added. Is that safe to use? Do we have any notes specifying the changes?

**Comment by adolski** [comment]
@mannu2612 Image API v3 is available on the `develop` branch. Define "safe." :smile:

Are you looking for [CHANGES.md](https://github.com/cantaloupe-project/cantaloupe/blob/develop/CHANGES.md) & [UPGRADING.md](https://github.com/cantaloupe-project/cantaloupe/blob/develop/UPGRADING.md)?

**Comment by ksclarke** [comment]
> @ksclarke Were you able to solve this?

@mannu2612 , define "solve" :smile:

Deleting the image from the derivatives cache has gotten us past that one problem, and we've not seen the issue pop back up again. It doesn't make sense to me why that would have solved the problem though.

**Comment by mannu2612** [comment]
> @mannu2612 Image API v3 is available on the `develop` branch. Define "safe." 😄
> 
> Are you looking for [CHANGES.md](https://github.com/cantaloupe-project/cantaloupe/blob/develop/CHANGES.md) & [UPGRADING.md](https://github.com/cantaloupe-project/cantaloupe/blob/develop/UPGRADING.md)?

Since we haven't done a major release to v5 yet, I was wondering if there's any open issue that is preventing this release. Let me try 5.0.0-SNAPSHOT and verify. Thanks for pointing me to upgrades notes :)

**Comment by mannu2612** [comment]
> > @ksclarke Were you able to solve this?
> 
> @mannu2612 , define "solve" 😄
> 
> Deleting the image from the derivatives cache has gotten us past that problem, and we've not seen the issue pop back up again. It doesn't make sense to me, given what I saw in the logs, why that would have solved the problem though. _* shrug *_

@ksclarke I need derivative S3 cache to improve the tile load performance in a web application. I make 40 different tile requests in parallel for the same jp2 image to pre-convert them to jpg and cache. These requests work fine with the cache disabled but crashes when I enable it. I hoped for a "solution" where cache works :sweat_smile: . I will try 5.0.0-SNAPSHOT to see whether upgraded Kakadu version helps, else I'd be very thankful if you can direct me to debug this issue and contribute.

**Comment by ksclarke** [comment]
Just in case I wasn't clear, in our case clearing the cache was a one-time fix. We were able to re-enable it again and it worked from then on; so it seems like our problem involved a bad derivative image that got into the cache somehow (rather than a persistent issue with the caching mechanism itself). Once we removed that image from the cache it's been working fine since (including for that same image again -- so in this/our case, it wasn't an issue with Kakadu being unable to process the image).

Also, fwiw, I'm not speaking on behalf of the project. Despite the "contributor" next to my name (I think I made a very trivial PR here once long ago), I'm just another user of cantaloupe. :-)

**Comment by cmhdave** [comment]
Well I've been bit by this bug today while testing our latest upgraded Kakadu processor.  We are using v8.0.6 of the Kakadu processor and the latest 5.0.0-SNAPSHOT branch of Cantaloupe. I can get this to fail fairly consistently in our dev environment.  The EASIEST way for me to replicate this issue is by trying to load full size images that are 10000x13000.  I have a Jmeter test that loops through ~20 images with 10 consecutive users.  The requests are always /full/full/0/default.jpg so it's pretty memory intensive.  After about 1 minute the server crashes with this error.  We handle extremely high volumes of requests so 10 consecutive users is not even a realistic production scenario. 

I was able to mitigate the issue slightly by restricting the allowed size to just 9000x9000 (max_pixels = 81000000) but all that really does is just eliminates the problem images from even being rendered instead of solving the problem.

I will see if I can package up the images tomorrow and attach them here.

**Comment by cmhdave** [comment]
A couple more details from a sample run this morning on a single very large 10000x13000 image at 600dpi. I ran it in  JMeter with 10 consecutive users all requesting the same image.  The VM I'm running on is very simple, 2 CPUs, 4GB ram.  The JVM is set to use 2GB ram.  The image file size is 3MB

I get the below in the log 10 times consecutively
```
o07:08:59 07:08:59.350 [qtp1658520865-45] ESC[39mDEBUGESC[0;39m ESC[36me.i.l.c.c.FilesystemCacheESC[0;39m - newDerivativeImageInputStream(): /tmp/cantaloupe/cache/image/74/c7/27/74c727b4dcf96d72e3a9c4ef01cdf54c_b8d30f83ec1dc6886d60784212be062f.jpg
o07:08:59 07:08:59.352 [qtp1658520865-45] ESC[39mDEBUGESC[0;39m ESC[36me.i.l.c.c.FilesystemCacheESC[0;39m - newOutputStream(): miss; caching alltypes:30_1:1_encode:jpg_JPEG_80_interlace_#FFFFFF_8_e95198c48292fa4c8fc0617e850c1f6d
```
then I get these two log statements 10 times consecutively
```
o07:08:59 07:08:59.390 [qtp1658520865-45] ESC[39mDEBUGESC[0;39m ESC[36me.i.l.c.r.ImageRepresentationESC[0;39m - Writing to the response & derivative cache simultaneously
o07:08:59 07:08:59.400 [qtp1658520865-45] ESC[39mDEBUGESC[0;39m ESC[36me.i.l.c.p.c.j.JPEG2000KakaduImageReaderESC[0;39m - Rendered region 0,0/10200x13200; source 0,0/10200x13200; 0x reduction factor; differential scale 10000000/10000000; PPI 600.0x600.0
```
Then I get a Java heap space error.  Then I get this error 10 consecutive times in the log

```
e07:09:08 Error in Kakadu Region Decompressor:
e07:09:08 The `kdu_region_decompressor::get_rendered_image_dims' function should not be
e07:09:08 called with a `codestream' argument between calls to
e07:09:08 `kdu_region_decompressor::start' and `kdu_region_decompressor::finish'.
```

So this might be a completely separate issue as I'm just testing against a single file this time and not 20 different files.  I wanted to include the details though since the image is from the same collection.

**Comment by cmhdave** [comment]
I wasn't able to attach my example files to Github as the gzip file is ~45MB but here is a [Google Drive link](https://drive.google.com/file/d/1E9vsHVHlF0xuSiBpwX-EhUhWvDFLfeW4/view?usp=sharing) to it.

In that file you'll see about 16-17 JP2 images, a sample.csv file and a kakadu.jmx file which is the Jmeter test plan I used.  You'll need to update the location of the CSV file in the Jmeter test plan under "Thread Group -> CSV Data Set Config"

Here is my java command line for running Cantaloupe

```
java -Xmx2g -Dcantaloupe.config=cantaloupe.properties -jar cantaloupe-5.0-SNAPSHOT.jar
```
and here is the JVM I'm using (Oracle 11.0.7)
```
java version "11.0.7" 2020-04-14 LTS
Java(TM) SE Runtime Environment 18.9 (build 11.0.7+8-LTS)
Java HotSpot(TM) 64-Bit Server VM 18.9 (build 11.0.7+8-LTS, mixed mode)
```

**Comment by adolski** [comment]
Thanks @cmhdave! I can reproduce this using your jmeter script & images. It crashes 100% of the time within a few seconds when a derivative cache is enabled, but it runs and runs when the cache is disabled.

I still don't know what the problem is, but just being able to reproduce is a major milestone.

**Comment by adolski** [comment]
The problem is this line:

https://github.com/cantaloupe-project/cantaloupe/blob/6bc50ab0b0ec0fe8bb7d51684269685f1edf732f/src/main/java/edu/illinois/library/cantaloupe/resource/ImageRepresentation.java#L127

Comment that out and no more crashing.

That line works with all other processors, but KakaduNativeProcessor is unique because of all the native object setup and teardown that it requires. It may improve resilience in certain outlier cases, but removing it is certainly preferable to crashing. I need to look into the implications of removing it before I can call this fixed, though.

**Comment by adolski** [comment]
There is a fix for this now on `release/4.1` and `develop` branches. Thanks to everyone for helping to isolate and reproduce this.
